api/v1alpha3/kubeadm_control_plane_webhook.go

```diff
@@ -0,0 +1,137 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha3
+
+import (
+	"reflect"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+func (r *KubeadmControlPlane) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(r).
+		Complete()
+}
+
+// +kubebuilder:webhook:verbs=create;update,path=/mutate-cluster-x-k8s-io-v1alpha3-kubeadmcontrolplane,mutating=true,failurePolicy=fail,groups=cluster.x-k8s.io,resources=kubeadmcontrolplanes,versions=v1alpha3,name=default.kubeadmcontrolplane.cluster.x-k8s.io
+// +kubebuilder:webhook:verbs=create;update,path=/validate-cluster-x-k8s-io-v1alpha3-kubeadmcontrolplane,mutating=false,failurePolicy=fail,groups=cluster.x-k8s.io,resources=kubeadmcontrolplanes,versions=v1alpha3,name=validation.kubeadmcontrolplane.cluster.x-k8s.io
+
+var _ webhook.Defaulter = &KubeadmControlPlane{}
+var _ webhook.Validator = &KubeadmControlPlane{}
+
+// Default implements webhook.Defaulter so a webhook will be registered for the type
+func (r *KubeadmControlPlane) Default() {
+	if r.Spec.Replicas == nil {
+		replicas := int32(1)
+		r.Spec.Replicas = &replicas
+	}
+
+	if r.Spec.InfrastructureTemplate.Namespace == "" {
+		r.Spec.InfrastructureTemplate.Namespace = r.Namespace
+	}
+}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (r *KubeadmControlPlane) ValidateCreate() error {
+	var allErrs field.ErrorList
+
+	if r.Spec.Replicas == nil {
+		allErrs = append(
+			allErrs,
+			field.Required(
+				field.NewPath("spec", "replicas"),
+				"is required",
+			),
+		)
+	} else if *r.Spec.Replicas <= 0 {
```

Comment:
FYI, I don't think it's possible to ever get here... when I was testing, because the CRD has a scale subresource, there's code in the apiextensions-apiserver repo that checks this. I do think it's worth having this code here as a failsafe, but if you're testing and you don't see this error message, that's why.

Reply:
Even if the scale resource prevents negative numbers, we still need to guard against 0 here.

Reply:
To clarify, the presence of the scale subresource enables validation logic in the apiextensions-apiserver for create and update calls that yields this error (this is from a `kubectl create` call):

```
The KubeadmControlPlane "kcp" is invalid: .spec.replicas: Invalid value: -2: should be a non-negative integer
```

Reply:
It took me a while to find the code that emits `should be a non-negative integer` 😄

Reply:
I can update to using `==` instead of `<=` if that is preferred.

Reply:
No, this is fine. But it's worth recognizing we'll never hit this condition, and we might want a comment to that effect.

Reply:
Updated with a comment.